src/librustc_trait_selection/opaque_types.rs

```diff
@@ -716,24 +720,36 @@ where
             ty::Closure(_, ref substs) => {
                 // Skip lifetime parameters of the enclosing item(s)
 
-                for upvar_ty in substs.as_closure().upvar_tys() {
-                    upvar_ty.visit_with(self);
-                }
+                let ty = self.infcx.shallow_resolve(substs.as_closure().upvar_tuple_ty());
+                if let ty::Infer(ty::TyVar(_)) = ty.kind {
+                    // Not yet resolved.
+                    ty.super_visit_with(self);
+                } else {
+                    for upvar_ty in substs.as_closure().upvar_tys() {
+                        upvar_ty.visit_with(self);
+                    }
 
-                substs.as_closure().sig_as_fn_ptr_ty().visit_with(self);
+                    substs.as_closure().sig_as_fn_ptr_ty().visit_with(self);
+                }
             }
 
             ty::Generator(_, ref substs, _) => {
                 // Skip lifetime parameters of the enclosing item(s)
                 // Also skip the witness type, because that has no free regions.
 
-                for upvar_ty in substs.as_generator().upvar_tys() {
-                    upvar_ty.visit_with(self);
-                }
+                let ty = self.infcx.shallow_resolve(substs.as_generator().upvar_tuple_ty());
```

Comment:
As discussed, can't work because infer depends on middle

Reply:
in that case we could just add a method to the inference context, so that you wrote `self.infcx.upvar_tys(substs.as_generator())` (and make sure to note that the method exists from `upvar_tuple_ty`)